File: src/worker.mjs
/**
 * The lease loop — Drop's seam, extended (DESIGN §5.1):
 *  - live lane drained FIRST (interactive requests never queue behind
 *    scheduled polls); bulk long-polls only when live is empty;
 *  - every response body gzipped (SQS 256KB posture); post-compression
 *    overflow is a loud error result + metric, never a silent fallback;
 *  - 403s feed the circuit breaker; open breaker = no leasing;
 *  - the SQS visibility timeout is the lease: post result, then delete.
 */

import { gzipSync } from "node:zlib";
import { crPath } from "./cr-api.mjs";

const MAX_RESULT_BYTES = 250_000; // headroom under the 256KB SQS cap

// CR pacing (§5.2): the docs' ~2s guidance, and overage surfaces as 403.
// The scheduler budget caps the AVERAGE rate; this floor caps the
// INSTANTANEOUS rate — without it a queued burst rips at wire speed and
// trips the breaker (learned live, 2026-09-03: 92-job fan-out, 5x403).
const MIN_FETCH_INTERVAL_MS = 1500;

export function makeWorker({
  sqs, // { receive(queueUrl, waitSeconds), send(queueUrl, body), delete(queueUrl, receiptHandle) }
  queues, // { live, bulk, results }
  crFetch,
  breaker,
  gatewayId,
  gatewaySha = null, // checkout SHA; rides results so the fleet-version panel is real
  metrics = { fetchSucceeded() {}, overflow() {}, breakerOpen() {} },
  log = () => {},
  now = () => new Date(),
  sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms)),
}) {
  let lastFetchStartedAt = 0;

  async function pacedFetch(path) {
    const wait = lastFetchStartedAt + MIN_FETCH_INTERVAL_MS - now().getTime();
    if (wait > 0) await sleep(wait);
    lastFetchStartedAt = now().getTime();
    return crFetch(path);
  }
  function buildResult(job, fetched) {
    const base = {
      v: 1,
      job,
      gateway_id: gatewayId,
      ...(gatewaySha ? { gateway_sha: gatewaySha } : {}),
      fetched_at: now().toISOString(),
    };
    if (fetched.kind === "transport") {
      return {
        ...base,
        status: "error",
        error: { kind: "transport", message: fetched.message },
      };
    }
    if (fetched.status !== 200) {
      return {
        ...base,
        status: "error",
        http_status: fetched.status,
        error: { kind: "http", message: `HTTP ${fetched.status}` },
      };
    }
    const gz = gzipSync(Buffer.from(fetched.bodyText, "utf8"));
    const b64 = gz.toString("base64");
    if (b64.length > MAX_RESULT_BYTES) {
      metrics.overflow();
      return {
        ...base,
        status: "error",
        http_status: 200,
        error: {
          kind: "overflow",
          message: `gzipped body ${b64.length}B exceeds cap`,
        },
      };
    }
    return { ...base, status: "ok", http_status: 200, body_gzip_b64: b64 };
  }

  async function handleLease(queueUrl, message) {
    let job;
    try {
      job = JSON.parse(message.body);
      crPath(job); // throws on unknown endpoint before we spend a CR call
    } catch (err) {
      // Malformed job: never fetch; let it re-lease toward the DLQ.
      log("warn", `unleasable job: ${err.message}`);
      return { handled: false };
    }

    const fetched = await pacedFetch(crPath(job));
    if (fetched.kind === "http" && fetched.status === 403) {
      const opened = breaker.record403();
      if (opened) {
        metrics.breakerOpen();
        log("warn", "circuit breaker OPEN after consecutive 403s");
      }
    } else if (fetched.kind === "http" && fetched.status === 200) {
      breaker.recordSuccess();
    }

    const result = buildResult(job, fetched);
    await sqs.send(queues.results, JSON.stringify(result));
    await sqs.delete(queueUrl, message.receiptHandle);
    if (result.status === "ok") metrics.fetchSucceeded();
    return { handled: true, status: result.status, lane: job.lane };
  }

  /** One iteration: live first, then bulk. Returns what happened. */
  async function pollOnce() {
    if (breaker.isOpen()) {
      return { polled: "breaker_open" };
    }
    let message = await sqs.receive(queues.live, 1);
    let queueUrl = queues.live;
    if (!message) {
      message = await sqs.receive(queues.bulk, 10);
      queueUrl = queues.bulk;
    }
    if (!message) return { polled: "empty" };
    return { polled: "job", ...(await handleLease(queueUrl, message)) };
  }

  return { pollOnce, handleLease, buildResult };
}

File: internal/worker/worker.go
// Package worker: the lease loop — Drop's seam, extended (DESIGN §5.1).
//   - live lane drained FIRST; bulk long-polls only when live is empty
//     (bulk wait 4s keeps worst-case live pickup ~5s, inside the
//     server's 12s live window — the race observed live 2026-09-05);
//   - every response body gzipped (SQS 256KB posture); post-compression
//     overflow is a loud error result + metric, never a silent fallback;
//   - 403s feed the circuit breaker; open breaker = no leasing;
//   - the SQS visibility timeout is the lease: post result, then delete.
package worker

import (
	"bytes"
	"compress/gzip"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"

	"github.com/jthingelstad/elixir-mcp-collector/internal/breaker"
	"github.com/jthingelstad/elixir-mcp-collector/internal/crapi"
)

const (
	maxResultBytes = 250_000 // headroom under the 256KB SQS cap
	// The scheduler budget caps the AVERAGE rate; this floor caps the
	// INSTANTANEOUS rate — without it a queued burst rips at wire speed
	// and trips the breaker (learned live 2026-09-03: 92-job fan-out).
	minFetchIntervalMs = 1500
)

type Message struct {
	Body          string
	ReceiptHandle string
}

type SQS interface {
	Receive(ctx context.Context, queueURL string, waitSeconds int32) (*Message, error)
	Send(ctx context.Context, queueURL, body string) error
	Delete(ctx context.Context, queueURL, receiptHandle string) error
}

type Queues struct{ Live, Bulk, Results string }

type Metrics struct {
	FetchSucceeded func()
	Overflow       func()
	BreakerOpen    func()
}

type Worker struct {
	sqs        SQS
	queues     Queues
	fetch      func(ctx context.Context, path string) crapi.Result
	breaker    *breaker.Breaker
	gatewayID  string
	gatewaySha string
	metrics    Metrics
	log        func(level, msg string)
	now        func() time.Time
	sleep      func(d time.Duration)

	lastFetchStartedAt time.Time
}

type Config struct {
	SQS        SQS
	Queues     Queues
	Fetch      func(ctx context.Context, path string) crapi.Result
	Breaker    *breaker.Breaker
	GatewayID  string
	GatewaySha string
	Metrics    Metrics
	Log        func(level, msg string)
	Now        func() time.Time
	Sleep      func(d time.Duration)
}

func New(c Config) *Worker {
	if c.Now == nil {
		c.Now = time.Now
	}
	if c.Sleep == nil {
		c.Sleep = time.Sleep
	}
	if c.Log == nil {
		c.Log = func(string, string) {}
	}
	if c.Metrics.FetchSucceeded == nil {
		c.Metrics.FetchSucceeded = func() {}
	}
	if c.Metrics.Overflow == nil {
		c.Metrics.Overflow = func() {}
	}
	if c.Metrics.BreakerOpen == nil {
		c.Metrics.BreakerOpen = func() {}
	}
	return &Worker{
		sqs: c.SQS, queues: c.Queues, fetch: c.Fetch, breaker: c.Breaker,
		gatewayID: c.GatewayID, gatewaySha: c.GatewaySha, metrics: c.Metrics,
		log: c.Log, now: c.Now, sleep: c.Sleep,
	}
}

// resultEnvelope mirrors the Node worker's buildResult exactly (wire
// names canonical in packages/contracts; job passes through raw so the
// server sees the leased bytes unmodified).
type resultEnvelope struct {
	V          int             `json:"v"`
	Job        json.RawMessage `json:"job"`
	GatewayID  string          `json:"gateway_id"`
	GatewaySha string          `json:"gateway_sha,omitempty"`
	FetchedAt  string          `json:"fetched_at"`
	Status     string          `json:"status"`
	HTTPStatus *int            `json:"http_status,omitempty"`
	Error      *resultError    `json:"error,omitempty"`
	BodyGzipB6 string          `json:"body_gzip_b64,omitempty"`
}

type resultError struct {
	Kind    string `json:"kind"`
	Message string `json:"message"`
}

func isoMillis(t time.Time) string {
	return t.UTC().Format("2006-01-02T15:04:05.000Z")
}

func (w *Worker) buildResult(rawJob json.RawMessage, fetched crapi.Result) resultEnvelope {
	env := resultEnvelope{
		V: 1, Job: rawJob, GatewayID: w.gatewayID, GatewaySha: w.gatewaySha,
		FetchedAt: isoMillis(w.now()),
	}
	if fetched.Kind == "transport" {
		env.Status = "error"
		env.Error = &resultError{Kind: "transport", Message: fetched.Message}
		return env
	}
	status := fetched.Status
	env.HTTPStatus = &status
	if fetched.Status != 200 {
		env.Status = "error"
		env.Error = &resultError{Kind: "http", Message: fmt.Sprintf("HTTP %d", fetched.Status)}
		return env
	}
	var buf bytes.Buffer
	gz := gzip.NewWriter(&buf)
	_, _ = gz.Write([]byte(fetched.BodyText))
	_ = gz.Close()
	b64 := base64.StdEncoding.EncodeToString(buf.Bytes())
	if len(b64) > maxResultBytes {
		w.metrics.Overflow()
		env.Status = "error"
		env.Error = &resultError{Kind: "overflow", Message: fmt.Sprintf("gzipped body %dB exceeds cap", len(b64))}
		return env
	}
	env.Status = "ok"
	env.BodyGzipB6 = b64
	return env
}

func (w *Worker) pacedFetch(ctx context.Context, path string) crapi.Result {
	wait := w.lastFetchStartedAt.Add(minFetchIntervalMs * time.Millisecond).Sub(w.now())
	if wait > 0 {
		w.sleep(wait)
	}
	w.lastFetchStartedAt = w.now()
	return w.fetch(ctx, path)
}

type PollOutcome struct {
	Polled  string // "breaker_open" | "empty" | "job"
	Handled bool
	Status  string
	Lane    string
}

func (w *Worker) handleLease(ctx context.Context, queueURL string, m *Message) (PollOutcome, error) {
	var job crapi.Job
	raw := json.RawMessage(m.Body)
	path := ""
	if err := json.Unmarshal(raw, &job); err == nil {
		path, err = crapi.Path(job)
		if err != nil {
			w.log("warn", "unleasable job: "+err.Error())
			return PollOutcome{Polled: "job", Handled: false}, nil
		}
	} else {
		// Malformed job: never fetch; let it re-lease toward the DLQ.
		w.log("warn", "unleasable job: "+err.Error())
		return PollOutcome{Polled: "job", Handled: false}, nil
	}

	fetched := w.pacedFetch(ctx, path)
	if fetched.Kind == "http" && fetched.Status == 403 {
		if w.breaker.Record403() {
			w.metrics.BreakerOpen()
			w.log("warn", "circuit breaker OPEN after consecutive 403s")
		}
	} else if fetched.Kind == "http" && fetched.Status == 200 {
		w.breaker.RecordSuccess()
	}

	result := w.buildResult(raw, fetched)
	body, err := json.Marshal(result)
	if err != nil {
		return PollOutcome{}, err
	}
	if err := w.sqs.Send(ctx, w.queues.Results, string(body)); err != nil {
		return PollOutcome{}, err
	}
	if err := w.sqs.Delete(ctx, queueURL, m.ReceiptHandle); err != nil {
		return PollOutcome{}, err
	}
	if result.Status == "ok" {
		w.metrics.FetchSucceeded()
	}
	return PollOutcome{Polled: "job", Handled: true, Status: result.Status, Lane: job.Lane}, nil
}

// PollOnce: live first, then bulk — one iteration of the lease loop.
func (w *Worker) PollOnce(ctx context.Context) (PollOutcome, error) {
	if w.breaker.IsOpen() {
		return PollOutcome{Polled: "breaker_open"}, nil
	}
	m, err := w.sqs.Receive(ctx, w.queues.Live, 1)
	if err != nil {
		return PollOutcome{}, err
	}
	queueURL := w.queues.Live
	if m == nil {
		m, err = w.sqs.Receive(ctx, w.queues.Bulk, 4)
		if err != nil {
			return PollOutcome{}, err
		}
		queueURL = w.queues.Bulk
	}
	if m == nil {
		return PollOutcome{Polled: "empty"}, nil
	}
	return w.handleLease(ctx, queueURL, m)
}
